Docs/headless cms/sourcing from graphcms (#11655)
gatsbyjs/gatsby: docs/docs/sourcing-from-graphcms.md

One of the benefits of GraphCMS when working with Gatsby is that we support GraphQL natively which allows you to test your content queries before even starting your Gatsby project.

## Getting Started

In this guide we'll walk you through creating a complete project capable of querying data from GraphCMS.

### Install the Boilerplate

To begin, let's create a Gatsby starter site.

```bash
  # Optionally prefixed with `npx` if you don't have the gatsby cli installed.
  gatsby new gatsby-site https://github.com/gatsbyjs/gatsby-starter-default
```

Navigate inside of the project with `cd gatsby-site`.

### Add the Source Plugin

Additionally, you need the the `gatsby-source-graphql` library. Because GraphCMS uses GraphQL natively, we will take advantage of Gatsby's ability to simply stitch two GraphQL APIs together, reducing the time required to transform content. There is no need to use a special gatsby-source-x-cms plugin, the GraphQL source plugin is all you need.

You can install this component with:

```bash
  # Optionally with `npm install`

>   yarn add gatsby-source-graphql

Reviewer: We should show `npm install` over `yarn add` here.

```

### Configure the Plugin

> The last step required before you can query your data is to configure the `gatsby-source-graphql` plugin. Open `gatsby-config.js` and add the following object to the plugins array. This example uses an open API from GraphCMS but you will most likely want to replace this with your own API and provide a fieldName that makes the most sense to your project.

Reviewer: > This example uses an open API from GraphCMS but you will most likely want to replace this with your own API

Is there a link we could add for the reader to learn more about this?

```js
{
    resolve: "gatsby-source-graphql",
        options: {
        // The top level query type, can be anything you want!
        typeName: "GCMS",
        // The field you'll query against, can also be anything you want.
        fieldName: "gcms",
        // Your API endpoint, available from the dashboard and settings window.
        // You can use this endpoint that features US mountains for now.
        url: "https://api-euwest.graphcms.com/v1/cjm7tab4c04ro019omujh708u/master",
    },
},
```

If everything works correctly, you should now have your GraphCMS data added to the Gatsby source API!

### Querying for Content

> From the root of your project, run the development environment with `yarn run development`. Once the server has started and is error free, you should be able to open the following URL in your browser:

Reviewer: Same as above: we prefer showing `npm` over `yarn`.

> http://localhost:8000/___graphql

Reviewer 2: ```suggestion
`http://localhost:8000/___graphql`
```

> This will show you a "Graphiql" playground where you can test your new content API.

Reviewer 2: ```suggestion
This will show you an interface where you can test your new content API.
```

Playground is a different product, and we're potentially switching from GraphiQL to Playground, so keeping this generic avoids the need to update later.

Try running this query:

```js
query {
    gcms {
        mountains {
            title
            elevation
            image {
                url
            }
        }
    }
}
```

Again, if everything is working properly, you should see a successful response in the shape of:

```json
{
  "data": {
    "gcms": {
      "mountains": [
        {
          "title": "Denali",
          "elevation": 6190,
          "image": {

>             "url": "https://media.graphcms.com//J0rGKdjuSwCk7QrFxVDQ"

Reviewer 2: Is the double slash here intentional, or a copy-paste error?

Author: It's an artefact from File Stack.

          }
        },
        {
          "title": "Mount Elbert",
          "elevation": 4401,
          "image": {
            "url": "https://media.graphcms.com//JNonajrqT6SOyUKgC4L2"
          }
        }
        // ...more results
      ]
    }
  }
}
```

### Getting Content on the Page

> For the purpose of this tutorial I've removed all the layout, SEO, link or other special components that comprise a page in the Gatsby starter. Open the index file located at `src/pages/index.js` and replace the content with this code:

Reviewer 2: You could use `gatsby-starter-hello-world` instead, if you don't want to deal with any of the peripheral pieces of the default starter.

Author: I actually like leaving it in for people to have access to it as they'll likely want it for their projects and a lot of examples reference it. I just wanted to emphasize the relevant code. What do you think?

Author: Just adding a quick ping @jlengstorf

Reviewer 2: Maybe the wording is unclear, then? I thought you meant that you had installed the default starter, then deleted the code you weren't going to use. Do you mean that you're just omitting it in the code sample?

I think I'm unclear on the intention.

Author: Gotcha, I'll fix.

> For the purpose of this tutorial I've removed all the layout, SEO, link or other components that comprise a page in the Gatsby starter. The components are still there and 99% of users will likely want to put them back in once they understand what's happenig in the code. You are just looking at the nails for right now, but the hammers, saws and other tools are still in the toolbox. Open the index file located at `src/pages/index.js` and replace the content with this code:

Reviewer: Typo: `happening`. I like the tooling metaphor!

```js
import React from "react"
import { StaticQuery } from "gatsby"

const IndexPage = () => (
  <StaticQuery
    query={graphql`
      query {
        gcms {
          mountains {
            title
            elevation
          }
        }
      }
    `}
    render={data => (
      <div>
        <h1>USA Mountains</h1>
        <ul>
          {data.gcms.mountains.map(mountain => {
            const { title, elevation } = mountain
            return (
              <li>
                <strong>{title}:</strong>
                {elevation}
              </li>
            )
          })}
        </ul>
      </div>
    )}
  />
)

export default IndexPage
```

With this code, you have:

> 1. Added the "StaticQuery" component to a page that allows you to fetch content from the GraphQL API.

Reviewer 2: ```suggestion
1. Added the `StaticQuery` component to a page that allows you to fetch content from the GraphQL API.
```

2. Fetched some simplified data from the Mountain API, namely the title and elevation.

> 3. Rendered the list in the StaticQuery's RenderProp called "render".

Reviewer 2: ```suggestion
3. Rendered the list in the `StaticQuery`'s `render` prop.
```

Author: I'm ok with this. Would you not bother telling them that render uses the "render prop" pattern?

Reviewer 2: Unless you wanted to link to a definition of what "render props" are, I wouldn't worry about it here.

Author: 👍

## Summary

> Hopefully you've seen how easy it is to start working with GraphCMS and Gatsby. With projects of all sizes gravitating towards the benefits of the JAM stack, the time has never been better to learn how to work with Gatsby. Adding a content API in the back-end with GraphCMS provides a scalable CMS that you can start using within minutes and keep using for the life of your project.

Reviewer 2: ```suggestion
Hopefully you've seen how easy it is to start working with GraphCMS and Gatsby. With projects of all sizes gravitating towards the benefits of the JAMstack, the time has never been better to learn how to work with Gatsby. Adding a content API in the back-end with GraphCMS provides a scalable CMS that you can start using within minutes and keep using for the life of your project.
```

Author: So is JAMstack becoming an official term?

Reviewer 2: Yeah, it's got conferences, websites, etc. dedicated to it, and the branding has solidified around this spelling.

Author: 👍

> [Checkout GraphCMS today and build "fast websites", fast!](https://graphcms.com?ref="gatsby-headless-docs-bottom")

Reviewer: There should be two words in this usage: "Check out"
